components/ContainedInputViews/src/MDCContainedInputView.h

```diff
@@ -0,0 +1,319 @@
+    (MDCContainedInputViewState)containedInputViewState;
+
+/**
+ This method sets a color scheme for a given state.
+ */
+- (void)setContainedInputViewColorScheming:
+            (nonnull id<MDCContainedInputViewColorScheming>)containedInputViewColorScheming
+                                  forState:(MDCContainedInputViewState)textFieldState;
+
+/**
+ Returns the rect surrounding the main content, i.e. the area that the container should be drawn
+ around.
+ */
+@property(nonatomic, assign, readonly) CGRect containerFrame;
+
+/**
+ This API allows the user to override the default main content area height. The main content area is
+ the part of the view where the where the data input happens. It is located above the underline
+ label area. If this property is set to a value that's lower than the default main content area
+ height the value will be ignored in the calculation of the view's @c intrinsicContentSize.
+ */
+@property(nonatomic, assign) CGFloat preferredMainContentAreaHeight;
+
+/**
+ This API allows the user to override the default underline label area height. The underline label
+ area is the part of the view where the underline labels are. It is located below the main content
+ area. If this property is set to a value that's lower than the default underline label area height
+ the value will be ignored in the calculation of the view's @c intrinsicContentSize.
+ */
+@property(nonatomic, assign) CGFloat preferredUnderlineLabelAreaHeight;
+
+@end
+
+/**
+ This protocol represents a set of colors that are semantically meaningful and specific to
+ MDCContainedInputView. Each property corresponds to the color of one or more views that an
+ MDCContainedInputView manages at a given point of time.
+ */
+@protocol MDCContainedInputViewColorScheming <NSObject>
+/**
+ The color of the contained input view's text.
+ */
+@property(strong, nonatomic, readonly, nonnull) UIColor *textColor;
+/**
+ The color of the contained input view's underline label.
+ */
+@property(strong, nonatomic, readonly, nonnull) UIColor *underlineLabelColor;
+/**
+ The color of the contained input view's floating label.
+ */
+@property(strong, nonatomic, readonly, nonnull) UIColor *floatingLabelColor;
+/**
+ The color of the contained input view's placeholder label.
+ */
+@property(strong, nonatomic, readonly, nonnull) UIColor *placeholderColor;
+/**
+ The tint color of the contained input view's clear button.
+ */
+@property(strong, nonatomic, readonly, nonnull) UIColor *clearButtonTintColor;
+/**
+ The color the contained input view should apply during an error state.
+ */
+@property(strong, nonatomic, readonly, nonnull) UIColor *errorColor;
+@end
+
+/**
+ A base implementation of MDCContainedInputViewColorScheming. Intended to be subclassed for styles
+ that include additional elements that need to be colored.
+ */
+@interface MDCContainedInputViewColorScheme : NSObject <MDCContainedInputViewColorScheming>
```

Comment:
Is it the fact that that it has the word "scheme" in it, or is the fact that it's a data structure comprised of colors? Which is the bad thing? If it's just the word "scheme" that's problematic, I'd be happy to change it, perhaps to "view model." That's really what it is. It's less about "primaryColor" and "secondaryColor" and more about "underlineLabelColor" and "floatingLabelColor," if that makes sense. Sort of a subtle difference.

I took great effort to ensure that the color scheming is in line with our conventions from the clients' perspective. They pass container schemes into methods like `-applyOutlinedThemeWithScheme:` and the view is themed.

Currently this stuff is crucial to how subviews and sublayers are colored, so getting rid of it would require that some other mechanism take its place.

Maybe it would be less scary if I explained how it came to be. Initially, as you might recall, the SimpleTextField owned a container scheme. Because it always had it on hand, it could look at it during every layout pass, adapt the color scheme into a set of colors for its subviews, and then apply those colors if necessary. Making the shift away from constant MDCColorScheme access to fleeting exposure during a theming extension method required that I take these state dependent combinations of colors and save them for later. What you see here is the method I chose for doing that. There are currently a lot of methods in our codebase like `-setBorderWidth:forState:`. I considered having a bunch of methods like `-setFloatingLabelColor:forState:` and `-setUnderlineLabelColor:forState:`, but decided it would be better to just have one, like `-setColorViewModel:forState:`.

Reply:
https://github.com/material-components/material-components-ios/pull/6985/files/10faae53f1a24686ab6851be47ab7306d696d27f#r269345857 is addressing a more root question, so I'll wait to respond here until that thread is resolved.